posts/2023-06-12-streams-in-scala.adoc

```diff
@@ -0,0 +1,214 @@
+ }
+// And here is the type of parameter the newLL receives. It receives a call-by-name parameter!
+/** Creates a new LazyList. */
+@inline private def newLL[A](state: => State[A]): LazyList[A] = new LazyList[A](() => state)
+----
+
+This portion `( state: => State[A]  )` is called , call-by-name. The state parameter has a return type of => State[A].  This parameter is not evaluated when passed, it's only evaluated when a terminating method is called. So all transforming method operate on the state without it being called.
+
+The same CBN is used as in the case below:
+
+[source,scala]
+----
+LazyList.cons(1, LazyList.cons(2, LazyList.empty))
+----
+
+The parameters below are called call-by name
+
+[source,scala]
+----
+/** An alternative way of building and matching lazy lists using LazyList.cons(hd, tl).
+  */
+object cons {
+  /** A lazy list consisting of a given first element and remaining elements
+    *  @param hd   The first element of the result lazy list
+    *  @param tl   The remaining elements of the result lazy list
+    */
+  def apply[A](hd: => A, tl: => LazyList[A]): LazyList[A] = newLL(sCons(hd, newLL(tl.state)))
+
+  /** Maps a lazy list to its head and tail */
+  def unapply[A](xs: LazyList[A]): Option[(A, LazyList[A])] = #::.unapply(xs)
+}
+----
+
+[[lazylist-implementation]]
+== A Simple Use-case of Scala Stream
+Let's consider a real-life scenario: Assume we are tasked with finding specific terms (e.g., success, failure, etc.) within large log files from various services running on our server. Our objective is to retrieve all occurrences of these terms and have the option to select the first few results. As you may have known, Scala provides a Source API for reading files. In this task, we would compare two approaches and see why one is better than the other.
+
+=== 1st Attempt:
+[source,scala]
+----
+files.map {
+ case (file) =>
+ Source.fromFile(file).getLines().toList.filter(_.contains("Success"))
+  .take(10)
+}
+----
+In the above snippet,we chained getLines and toList which ends up loading the content of the file to memory before filter is called. When we call toList, it evaluates immediately, and only after having the whole data, the filtering is applied. using a strict data structure like List would be a bad idea because of memory usage because the file could be large.
+
+=== 2nd Attempt:
+[source,scala]
+----
+files.map {
+ case (file) => Source.fromFile(file).getLines().to(LazyList).filter(_.contains("Success")).take(10)
+
+}
+----
+In the above snippet, we chained getLines with to(LazyList). With this, the content of the files is not loaded into the memory we then apply the filter and take functions which still don't load the content. The content of the file is only loaded when we call a terminating method like foreach. So the benefit of this is that it would only compute the first ten lines that match the filter predicate so that we don't end up loading everything from file.
```

Comment:
```suggestion
In the above snippet we chained getLines with to(LazyList). With this, the content of the files is not loaded into memory.  We then apply the filter and take functions which still don't load the content. The content of the file is only loaded when we call a terminating method like foreach. So the benefit of this is that it would only compute the first ten lines that match the filter predicate so that we don't end up loading everything from file.
```